### ISSUE TITLE
Integration of ROM Hacks into Online Updater

### ISSUE THREAD

**Issue by SpaceAgeHero** [libretro/RetroArch#8472]
Dear devs,

I would like to ask for an integration of Rom Hacks (preferably romhacking.net contents) into the online updater of RetroArch.

Here are some additional thoughts:
- maybe somehow like cheats work?
- download complete hack-sets system-wise
- RomHacking.net propably doesn't provide an API so contents must be mirrored / hosted somewhere else?
- ability to enable a specific Rom Hack on demand for soft-patching

**Comment by orbea** [comment]
With all due honesty this sounds like a huge maintenance burden and bug farm with very little benefit for most users.

**Comment by hizzlekizzle** [comment]
I think it's a great idea, but yes, without an API from RHDN, it would be a maintenance burden, for sure.

As for being of little benefit to most users, I think that's what OP is getting at: it would make romhacks more accessible to people who know nothing about them.

**Comment by orbea** [comment]
> As for being of little benefit to most users, I think that's what OP is getting at: it would make romhacks more accessible to people who know nothing about them.

I see what you mean, but do most people even want to use hacks? My impression is that its a niche use case, but maybe I'm mistaken?

I think the biggest problems excluding the lack of an API is making sure the content and patch matches, romhacking is full of patches that only work with very specific roms, some which depend on other hacks or changing the rom from headered and back and forth... (i.e the first patch needs a headered rom and the second patch needs the first patch and a unheadered rom)

I would think this is out of scope, but then I'm not sure we have a clearly defined scope in the first place...

**Comment by i30817** [comment]
This will probably never happen, but just in case, a aside:

The ips rom hacks for pc-98 and snes that 'require' a header can be made to apply on roms without a header with https://github.com/heuripedes/ipsbehead/releases

bps romhacks of course, never need to care about headers.


btw, this would get significantly more complicated for iso romhacks. Some of them need conversion to iso first, some of them need cdmage image rebuilding etc.

**Comment by i30817** [comment]
I also had the idea of using the urls on the romhack dats on liibretro-database to allow a person to see what 'romhacks' entries are applicable to a original game and opening the urls from retroarch.

I believe that i gave up on that when i figured out that retroarch s currently not even using the CRC32 of roms to recognize romhacks.

**Comment by hizzlekizzle** [comment]
@i30817 good point on the lack of standards for patch containers. Certainly complicates things.

@orbea I would think the fact that we do CRC scanning would help with the header thing, but it would really require the hacks be pre-sorted to know what goes with what.

**Comment by RobLoach** [comment]
[Softpatching in RetroArch](https://docs.libretro.com/guides/softpatching/) works. Download the patch, and name it the same as your rom file, in the same directory, launch the game. RetroArch will apply the patch in memory.

- Having a UI around ROMHacking.net to download patches would be neat, but as orbea mentioned, it would be a lot of work to maintain.
- i30817 has submitted a few of the most popular hacks to [libretro-database/metadat/hacks](https://github.com/libretro/libretro-database/tree/master/metadat/hacks).
- The Menu UI is difficult to work with, but I could imagine providing users a way to run roms with a selected patch. Have RetroArch soft-patch with the selected file...
    ![patchui](https://user-images.githubusercontent.com/25086/54443471-0c700a80-4717-11e9-9457-b02f4b451246.png)

**Comment by i30817** [comment]
Unfortunately softpatching is not suitable for roms larger than 32mb or so. Retroarch doesn't even support it for the n64 (or NDS) much less the ps1 and larger.

It's a elegant idea, don't get me wrong (to calculate the crc32  of ips/bps files instead of the final result to match it to a romhack entry).

However, people use different patch formats than the original for valid reasons all the time (ipsbehead is a example, but not the only one).

**Comment by orbea** [comment]
> I would think the fact that we do CRC scanning would help with the header thing, but it would really require the hacks be pre-sorted to know what goes with what.

Exactly, also consider that new hacks and translations are added frequently, often not in English. :)

**Comment by i30817** [comment]
That already occurs. Having a incomplete database is not enough reason to give up having a database, and if people are especially attached to, say, spanish translations, they can PR the database using rhdndat to keep the patches up to date, like the english ones.

**Comment by orbea** [comment]
> if people are especially attached to, say, spanish translations, they can PR the database using rhdndat to keep the patches up to date, like I do for many english ones (not all ofc).

Are you volunteering?

**Comment by i30817** [comment]
No, because i don't know spanish and therefore have no interest in spanish translations (in spite of living right next to spain). I'm just saying that the idea that 'we don't want to recognize translations and hacks because some aren't popular enough to attract a PRs' is bizarre.

**Comment by orbea** [comment]
You're not understanding my point, the point is that this is massive ongoing project with no willing maintainers.

**Comment by i30817** [comment]
This - to be clear, just recognizing romhacks, *not* downloading them - is a *optional* effort that needs no ongoing maintainers in the main team. If that's your attitude, you might as well gut everything not from dumper groups from libretro-database.

**Comment by retro-wertz** [comment]
whats difference of just browsing to romhacking sites like romhacking, download patch and rename as needed? also, how to know if the patch requires a patched rom or not? who is gonna maintain database for this (ugh.. more database files....)

**Comment by Lreaper** [comment]
It would be really neat if there was a menu option to load content with a specific ips patch, I'm aware softpatching already works if you have the patch on the same folder as the rom file but what if you have more than one patch for one specific game? Having a menu option to load content+patch would allow the user to have several romhacks available for any given game without having to constantly rename/move patches around.

**Comment by i30817** [comment]
That would require ips/bps softpatching to be externalized to cores, which is something that RA-libretro has steadfastly refused to do for many years.

Though i'm sympathetic about the fact that if they did it would be done badly (by using a tmp file hardpatch) because Windows filesystem / ramdisk apis utterly suck and RA is not willing to make a feature only work on unix or whatever. Fuck it.

**Comment by Mte90** [comment]
I agree that a system to download is not easy but can be a step 2, better focusing on reconizing rom hacks more easily.
Right now softpatching seems cool but what to do in case that the patch are more then one?
Or in the case that the format is not supported by the softpatching?
Also right now renaming the patch doesn't let us to remember watch was the patch doing.

An user case like mine:

* Get the rom
* Download different patch
* Try them one by one
* See if it is possible to merge more then one

Right now retroarch use crc that is an issue on detecting some roms so maybe we can find another way to add roms already patched.
Maybe adding in the rom file "Amazing-game.nocrc.zip" and if the filename include `nocrc` doesn't use the rom scan but add to the playlist of the console (reconizing the file format content).

Another solution can be simplify the adding of patches to https://github.com/libretro/libretro-database/tree/master/dat from the community so the patch are availabl right there so the scanning is more simple.

**Comment by i30817** [comment]
>  Right now softpatching seems cool but what to do in case that the patch are more then one?

Modern filesystems can deduplicate roms. When i need this I just create a hardlink, but i bet that some modern filesystems are thinking of doing this automatically.

> Right now retroarch use crc that is an issue on detecting some roms so maybe we can find another way to add roms already patched.

Merged patches willl probably never be 'automatically' detected. It's simply impossible with any checksum approach. I do put in some merged patches into libretro-database but that's because i merged them manually as good combinations, but even then i don't bother with edgy 'patchers with 10.000 options' that some hackers are doing.

> Right now retroarch **[don't]** use crc **[except on netplay, only that crc is of the game executable so mostly useless for the current database of hacks]** that is an issue on detecting some roms so maybe we can find another way to add roms already patched.

FTFY. This *regression* was added because people were complaining that their self-dumped games in random formats were not being recognized by the scanner and were slow to scan so now  your 'reward' is that no game hack is scanning. To solve this All Roads Lead to [customization of the scanner](https://github.com/libretro/RetroArch/issues/8672).

A issue that all RA contributors are not interested in btw.

> Another solution can be simplify the adding of patches to https://github.com/libretro/libretro-database/tree/master/dat from the community so the patch are availabl right there so the scanning is more simple.

No way. You probably don't realize this but CD patching of 'modern' games is absolutely insane in required space (because modern games stuff everything in compressed huge containers where a binary change earlier in the stream modifies everything else.

**Comment by Imrhien** [comment]
All these ideas are cool but definitely hard to implement.

Short-term - why not implement a way that the RA romscanner can still add unrecognised games to the playlist?

I'd be happy with RA just parsing in the filename of the unrecognised ROMs and adding them to the playlist. Perhaps all the unrecognised titles go into a separate playlist (as RA would have no way to know which system to associate them with).

**Comment by i30817** [comment]
It has that, it's called 'the manual scanner'.

If you want images with your games, might i recommend my own tool that uses fuzzy matching (of the game name) to attempt to match it to the libretro database filenames?

https://github.com/i30817/libretrofuzz

**Comment by i30817** [comment]
Btw the situation is slightly better today because a outside contributor coded the softpatch part of RA to accept multiple patches at once, so now if you want that, you don't need to create a merged patch (tiresome), just add '.ips1, .ips2' after the main one.

Well, except in the unfortunately common cases where the patch has some special needs for the source rom, like a header or a iso instead of the standard redump cue dump etc.

btw for snes and pce headers, this utility is useful https://github.com/heuripedes/ipsbehead

And works almost all the time (if it doesn't work it's because the original ips was doing something wrong and fucking with the header somehow).

**Comment by mdsmendes94** [comment]
That would be an amazing feature to have for sure, RetroArch integration with RHDN or MetroidConstruction or any other host that provides romhacks

**Comment by romh-acking** [comment]
Someone on romhack.ing (a romhacking.net spiritual successor) brought up the possibility of a RetroArch integration [here](https://romhack.ing/forum/topic/KcYHg5YBOY1av0NOfDTv).

**Comment by RobLoach** [comment]
A ROM Hack Manager core would be a cool idea. While not the best user experience, being able to run a core, and have it manage which .ips files are active is, after development, feasible. Pretty awesome for romhack.ing to set that up. Thanks for reaching out.
